ui/raidboss/data/06-ew/raid/p12s.ts

```diff
@@ -604,30 +642,191 @@ const triggerSet: TriggerSet<Data> = {
+    // The people in the tower find the people who have the opposite attribute to the debuff and put them in four places.
+    // At NE NW SE SW as a # shape. The position of outside Anthropos is fixed by two situation.
+    // {[97, 75], [125, 97], [103, 125], [75, 103]} and {[103, 75], [125, 103], [97, 125], [75, 97]}. The Anthropos will cast
+    // 'Searing Radiance' for light beam and 'Shadowsear' for dark beam. We use those as a trigger for Tower players place
+    // the Tower.
     // When debuffs expire and towers drop, their debuff changes to lightTilt or darkTilt (same as tower color).
     // At the same time the towers drop, the 4 tethered players receive lightTilt or darkTilt depending on their tether color.
+    //
+    {
+      id: 'P12S Engravement 1 Tether Tracker',
+      type: 'Tether',
+      netRegex: { id: Object.keys(anthroposTetherMap), source: 'Anthropos' },
+      run: (data, matches) => {
+        const tetherType = anthroposTetherMap[matches.id];
+        if (tetherType === undefined)
+          return;
+        data.engravement1TetherPlayers[matches.sourceId] = tetherType;
+        data.engravement1TetherIds.push(parseInt(matches.sourceId, 16));
+      },
+    },
+    {
+      id: 'P12S Engravement 1 Beam',
+      type: 'StartsUsing',
+      netRegex: { id: Object.keys(tetherAbilityToTowerMap), source: 'Anthropos' },
+      condition: (data) => data.engravementCounter === 1,
+      alertText: (data, matches, output) => {
+        if (data.me === matches.target) {
+          if (matches.id === '82F1')
+            return output.lightBeam!();
+          return output.darkBeam!();
+        }
+      },
+      outputStrings: {
+        lightBeam: {
+          en: 'light beam',
+          cn: '引导光激光',
+        },
+        darkBeam: {
+          en: 'dark beam',
+          cn: '引导暗激光',
+        },
+      },
+    },
     {
       id: 'P12S Engravement 1 Tower Drop',
       type: 'GainsEffect',
       netRegex: { effectId: engravementTowerIds },
       condition: (data) => data.engravementCounter === 1,
       durationSeconds: (_data, matches) => parseFloat(matches.duration),
+      promise: async (data) => {
+        data.combatantData = [];
+        data.combatantData = (await callOverlayHandler({
+          call: 'getCombatants',
+          ids: data.engravement1TetherIds,
+        })).combatants;
+      },
       alertText: (data, matches, output) => {
         data.engravement1Towers.push(matches.target);
+
+        for (let index = 0; index < data.engravement1TetherIds.length; index++) {
+          const x = data.combatantData[index]?.PosX;
+          if (x === undefined)
+            return;
+          const y = data.combatantData[index]?.PosY;
+          if (y === undefined)
+            return;
+
+          const ids = data.engravement1TetherIds[index];
+          if (ids === undefined)
+            return;
+
+          const tempColor = data.engravement1TetherPlayers[ids.toString(16).toUpperCase()];
```

Comment:
```suggestion
        for (const combatant of data.combatantData) {
          const x = combatant.PosX;
          const y = combatant.PosY;
          const combatantId = combatant.ID;
          if (combatantId === undefined)
            return;

          const tempColor = data.engravement1TetherPlayers[combatantId.toString(16).toUpperCase()];
```

`data.combatantData` is (unfortunately) not sorted in the order as the ids provided, and so you can't use the same index to figure out the color.